File: src/gpu/command_pool.cc
#include "command_pool.h"

#include <vulkan/vulkan_core.h>

#include "status.h"

using namespace villa;

void CommandPool::init(VkDevice device, uint32_t graphics_queue_family) {
   device_ = device;

   VkCommandPoolCreateInfo create_info = {
       .sType = VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO,
       .flags = VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT,
       .queueFamilyIndex = graphics_queue_family,
   };
   VILLA_VK(vkCreateCommandPool(device, &create_info, nullptr, &command_pool_));
}

void CommandPool::deinit() {
   if (command_pool_ != VK_NULL_HANDLE) {
      vkDestroyCommandPool(device_, command_pool_, nullptr);
   }
}

VkCommandBuffer CommandPool::allocate() {
   VkCommandBufferAllocateInfo alloc_info = {
       .sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO,
       .commandPool = command_pool_,
       .level = VK_COMMAND_BUFFER_LEVEL_PRIMARY,
       .commandBufferCount = 1,
   };

   VkCommandBuffer result;
   VILLA_VK(vkAllocateCommandBuffers(device_, &alloc_info, &result));

   return result;
}

File: runtime/gpu/vulkan/command_pool.zig
const std = @import("std");
const builtin = @import("builtin");
const build_options = @import("build_options");
const vk = @import("vulkan");
const vkAssert = @import("status.zig").vkAssert;

pub const CommandPool = struct {
    device: vk.VkDevice,
    command_pool: vk.VkCommandPool,

    const Self = @This();

    pub fn init(device: vk.VkDevice, graphics_queue_family: u32) Self {
        const create_info = vk.VkCommandPoolCreateInfo{
            .sType = vk.VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO,
            .flags = vk.VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT,
            .queueFamilyIndex = graphics_queue_family,
        };
        var command_pool: vk.VkCommandPool = null;
        vk.vkCreateCommandPool(device, &create_info, null, &command_pool);

        return .{ .device = device, .command_pool = command_pool };
    }

    pub fn allocate(self: *Self) error{VulkanFailure}!vk.VkCommandBuffer {
        const buffer_alloc_info = vk.VkCommandBufferAllocateInfo{
            .sType = vk.VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO,
            .commandBufferCount = 1,
            .commandPool = self.command_pool,
            .level = vk.VK_COMMAND_BUFFER_LEVEL_PRIMARY,
        };
        var command_buffer: vk.VkCommandBuffer = null;
        try vkAssert(vk.vkAllocateCommandBuffers(self.device, &buffer_alloc_info, &command_buffer));

        return command_buffer;
    }
};
